src/restic_fmt_plug.c

```diff
@@ -0,0 +1,334 @@
+	if (((p = strtokm(NULL, "*")) == NULL) || !isdec(p)) /* scrypt: N */
+		goto err;
+	if (((p = strtokm(NULL, "*")) == NULL) || !isdec(p)) /* scrypt: r */
+		goto err;
+	if (((p = strtokm(NULL, "*")) == NULL) || !isdec(p)) /* scrypt: p */
+		goto err;
+	if (((p = strtokm(NULL, "*")) == NULL) || !ishexn(p, 64)) /* scrypt: salt */
+		goto err;
+	if (((p = strtokm(NULL, "*")) == NULL) || !ishexn(p, 160)) /* restic: data */
+		goto err;
+
+	MEM_FREE(keeptr);
+	return 1;
+err:
+	MEM_FREE(keeptr);
+	return 0;
+}
+
+void *restic_get_binary(char *ciphertext)
+{
+	static unsigned char c[BINARY_SIZE];
+	char *p;
+	int i;
+	p = strrchr(ciphertext, '*') + 1;
+	// Restic design doc: computed MAC will not match the last 16 bytes of the data
+	p += (160 - 16) * 2;
+	for (i = 0; i < BINARY_SIZE; i++) {
+		c[i] = (atoi16[ARCH_INDEX(*p)] << 4) | atoi16[ARCH_INDEX(p[1])];
+		p += 2;
+	}
+	return c;
+}
+
+void *restic_get_salt(char *ciphertext)
+{
+	char *ctcopy = strdup(ciphertext);
+	char *keeptr = ctcopy;
+	int i;
+	char *p;
+	restic_salt *cur_salt;
+
+	cur_salt = mem_calloc_tiny(sizeof(restic_salt), MEM_ALIGN_SIMD);
+
+	ctcopy += TAG_LENGTH;
+
+	strtokm(ctcopy, "*"); /* SCRYPT */
+
+	p = strtokm(NULL, "*");
+	cur_salt->N = atoi(p);
+
+	p = strtokm(NULL, "*");
+	cur_salt->r = atoi(p);
+
+	p = strtokm(NULL, "*");
+	cur_salt->p = atoi(p);
+
+	p = strtokm(NULL, "*");
+	for (i = 0; i < sizeof(cur_salt->salt); i++)
+		cur_salt->salt[i] = atoi16[ARCH_INDEX(p[i * 2])] * 16 + atoi16[ARCH_INDEX(p[i * 2 + 1])];
+
+	p = strtokm(NULL, "*");
+	for (i = 0; i < sizeof(cur_salt->data); i++)
+		cur_salt->data[i] = atoi16[ARCH_INDEX(p[i * 2])] * 16 + atoi16[ARCH_INDEX(p[i * 2 + 1])];
+
+	MEM_FREE(keeptr);
+	return cur_salt;
+}
+
+unsigned int restic_tunable_cost_N(void *salt)
+{
+	restic_salt *rs = salt;
+	return (unsigned int)rs->N;
+}
+
+unsigned int restic_tunable_cost_r(void *salt)
+{
+	restic_salt *rs = salt;
+	return (unsigned int)rs->r;
+}
+
+unsigned int restic_tunable_cost_p(void *salt)
+{
+	restic_salt *rs = salt;
+	return (unsigned int)rs->p;
+}
+
+void restic_set_salt(void *salt)
+{
+	cur_salt = (restic_salt *)salt;
+}
+
+void restic_set_key(char *key, int index)
+{
+	strnzcpy(saved_key[index], key, sizeof(*saved_key));
+}
+char *restic_get_key(int index)
+{
+	return saved_key[index];
+}
+
+int restic_cmp_all(void *binary, int count)
+{
+	int index;
+
+	for (index = 0; index < count; index++)
+		if (((uint32_t *)binary)[0] == crypt_out[index][0])
+			return 1;
+	return 0;
+}
+
+int restic_cmp_one(void *binary, int index)
+{
+	return !memcmp(binary, crypt_out[index], BINARY_SIZE);
+}
+
+int restic_cmp_exact(char *source, int index)
+{
+	return 1;
+}
+
+static int restic_crypt_all(int *pcount, struct db_salt *salt)
+{
+	int count = *pcount;
+	int index;
+	int failed = 0;
+	yescrypt_params_t params = {.N = cur_salt->N, .r = cur_salt->r, .p = cur_salt->p};
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+	for (index = 0; index < count; index += MIN_KEYS_PER_CRYPT) {
```

Comment:
This code appears to assume that `MIN_KEYS_PER_CRYPT` is 1, and will misbehave otherwise. So the loop should use simple `index++`, which would eliminate such assumption.

Reply:
thank you again for your explanations: in the beginning not all parameters
were clear to me. Also this part I had just taken over without greater
thinking from `ethereum_fmt_plug.c`.